### PR TITLE
Support VIP

### PR DESCRIPTION
**Pull request by yaocw2020**
issue: https://github.com/harvester/harvester/issues/632

### CODE REVIEW

File: pkg/console/vip.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	lines := strings.Split(string(out), "\n")
+
+	var ip string
+	for _, line := range lines {
+		words := strings.FieldsFunc(line, func(c rune) bool {
+			if c == '=' || c == '\'' {
+				return true
+			}
+			return false
+		})
```
Comment (gitlawr) [inline pkg/console/vip.go:89]: We rely on wicked command output which is not readable from the code. Can you add some unit tests to indicate & ensure the output parsing works?

File: pkg/config/cos.go [hunk 2]
```diff
@@ -97,24 +99,44 @@ func ConvertToCOS(config *HarvesterConfig) (*yipSchema.YipConfig, error) {
...
+			initramfs.Files = append(initramfs.Files, yipSchema.File{
+				Path:        manifestsDirectory+canalConfig,
+				Content:     canalHelmChartConfig,
+				Permissions: 0600,
+				Owner:       0,
+				Group:       0,
+			})
+		}
+
+		logrus.Infof("aaa config: %+v", config)
```
Comment (gitlawr) [inline pkg/config/cos.go:124]: fix the log

**Comment by bk201** [comment]
We need to preload images for these pods:
Screen Shot 2021-09-07 at 4 13 15 PM

Maybe extracting them from the helm charts?

**Comment by bk201** [comment]
These new fields are mandatory:

```
	Vip       string `json:"vip,omitempty"`
	VipHwAddr string `json:"vipHwAddr,omitempty"`
	VipMode   string `json:"vipMode,omitempty"`
```

We need to:
- Add checks in https://github.com/harvester/harvester-installer/blob/master/pkg/console/validator.go
- Update doc for these new fields
   - https://github.com/harvester/docs/blob/v0.3-dev/docs/install/harvester-configuration.md
   - https://github.com/harvester/docs/blob/v0.3-dev/docs/install/pxe-boot-install.md
- Update config files in iPXE examples: https://github.com/harvester/ipxe-examples

**Comment by bk201** [comment]
Got an error on static method:
Screen Shot 2021-09-08 at 12 20 39 PM

**Comment by gitlawr** [comment]
We should display the VIP as the management URL post installation

File: pkg/console/dashboard_panels.go [hunk 3]
```diff
@@ -213,7 +213,7 @@ func syncManagementURL(ctx context.Context, g *gocui.Gui) {
 
 func doSyncManagementURL(g *gocui.Gui) {
 	managementURL := "Unavailable"
-	if managementIP := getFirstReadyMasterIP(); managementIP != "" {
+	if managementIP := getVIP(); managementIP != "" {
 		managementURL = fmt.Sprintf("https://%s:%s", managementIP, harvesterNodePort)
```
Comment (yaocw2020) [inline pkg/console/dashboard_panels.go:217]: No need harvesterNodePort.

File: pkg/console/install_panels.go [hunk 4]
```diff
@@ -1275,3 +1278,124 @@ func addUpgradePanel(c *Console) error {
...
+		closeThisPage()
+		return showNext(c, tokenPanel)
+	}
+	gotoVipPanel := func(g *gocui.Gui, v *gocui.View) error {
+		selected, err := askVipMethodV.GetData()
+		if err != nil {
+			return err
+		}
+		if selected == config.NetworkMethodDHCP {
+			spinner := NewSpinner(c.Gui, vipTextPanel, "Request IP through DHCP...")
```
Comment (yaocw2020) [inline pkg/console/install_panels.go:1315]: Requesting

File: pkg/console/validator.go [hunk 5]
```diff
@@ -102,6 +102,14 @@ func checkIP(addr string) error {
 	return nil
 }
 
+func checkHwAddr(hwAddr string) error {
+	if _, err := net.ParseMAC(hwAddr); err != nil {
+		return fmt.Errorf("%s is an invalid hardware address", hwAddr)
```
Comment (yaocw2020) [inline pkg/console/validator.go:107]: Add err message

File: pkg/console/vip.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+type vipAddr struct {
+	hwAddr   string
+	ipv4Addr string
+}
+
+func createMacvlan(name string) (netlink.Link, error) {
+	l, err := netlink.LinkByName(name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "fail to fetch %s", name)
```
Comment (yaocw2020) [inline pkg/console/vip.go:22]: failed

File: pkg/console/vip.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	macvlan := &netlink.Macvlan{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:        tempMacvlanName,
+			ParentIndex: l.Attrs().Index,
+		},
+	}
+
+	if err = netlink.LinkAdd(macvlan); err != nil {
+		return nil, errors.Wrapf(err, "fail to add %s", tempMacvlanName)
```
Comment (yaocw2020) [inline pkg/console/vip.go:32]: failed

File: pkg/console/vip.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	if err = netlink.LinkSetUp(macvlan); err != nil {
+		return nil, errors.Wrapf(err, "fail to set %s up", tempMacvlanName)
```
Comment (yaocw2020) [inline pkg/console/vip.go:35]: failed

File: pkg/console/vip.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	return netlink.LinkByName(tempMacvlanName)
+}
+
+func deleteMacvlan(l netlink.Link) error {
+	// It's necessary to set macvlan down at first to notify the wicked to clear the related files automatically.
+	if err := netlink.LinkSetDown(l); err != nil {
+		return errors.Wrapf(err, "fail to set %s down", err)
```
Comment (yaocw2020) [inline pkg/console/vip.go:44]: ditto

File: pkg/console/vip.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	if err := netlink.LinkDel(l); err != nil {
+		return errors.Wrapf(err, "fail to del %s", tempMacvlanName)
```
Comment (yaocw2020) [inline pkg/console/vip.go:47]: ditto

**Comment by yaocw2020** [comment]
> We need to preload images for these pods:
> Screen Shot 2021-09-07 at 4 13 15 PM
> 
> Maybe extracting them from the helm charts?

We will add the dependency charts information in the value.yaml file of the Harvester helm chart.

File: pkg/console/validator.go [hunk 11]
```diff
@@ -169,6 +196,10 @@ func (v ConfigValidator) Validate(cfg *config.HarvesterConfig) error {
 		return err
 	}
 
+	if err := checkVip(cfg.Vip, cfg.VipHwAddr, cfg.VipMode); err != nil {
```
Comment (bk201) [inline pkg/console/validator.go:199]: Should only check when `cfg.Install.Mode == config.ModeCreate`

Comment (yaocw2020) [reply pkg/console/validator.go:199]: Yes.

**Comment by bk201** [comment]
Update PXE example: https://github.com/harvester/ipxe-examples/pull/11/

**Review by gitlawr — APPROVED** [review]
LGTM

**Review by bk201 — APPROVED** [review]
LGTM

**Review by guangbochen — APPROVED** [review]

**Pull request merged**